### PR TITLE
Linux 5.11 compat

### PR DESCRIPTION
**Pull request by behlendorf**
### Motivation and Context

Issue #11387.  Linux 5.11 compat.

### Description

The Linux 5.11 kernel has made multiple changes to the block layer
interfaces which require some new compatibility code.  This PR contains
a stack of 5 patches each of which addresses a different block device
interface change.  An explanation of each commit is in the associated
commit comment.

* Linux 5.11 compat: lookup_bdev()
* Linux 5.11 compat: bio_start_io_acct() / bio_end_io_acct()
* Linux 5.11 compat: bdev_whole()
* Linux 5.11 compat: revalidate_disk_size()
* Linux 5.11 compat: blk_{un}register_region() 

### How Has This Been Tested?

Compiled on a 3.10 (CentOS 7) and unreleased 5.11 mainline kernel.
Ran the `sanity.run` ZTS tests to verify basic functionality.   The CI will
provide test coverage for a wide range of kernels and I'm planning to
run the full ZTS suite locally with the 5.11 kernel.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ckane** [comment]
Hey @behlendorf  - looks like on Dec 1 they made MODULE_LICENSE an error instead of a warn, which makes the kernel module build tests fail in `config/kernel.m4` because the `conftest.c` doesn't get those things. I modified the `ZFS_LINUX_TEST_PROGRAM` per below and I think that should get it. Going to test with both change sets merged later today:
```
dnl #
dnl # ZFS_LINUX_TEST_PROGRAM(C)([PROLOGUE], [BODY])
dnl #
m4_define([ZFS_LINUX_TEST_PROGRAM], [
$1
#include <linux/module.h>
int
main (void)
{
$2
        ;
        return 0;
};
MODULE_DESCRIPTION("conftest");
MODULE_AUTHOR(ZFS_META_AUTHOR);
MODULE_LICENSE(ZFS_META_LICENSE);
MODULE_VERSION(ZFS_META_VERSION "-" ZFS_META_RELEASE);
])
```

**Comment by ckane** [comment]
Also, I get the following error when running `configure` against the latest Linux git source tree:
```
checking whether PDE_DATA() is available... configure: error:
	*** None of the expected "PDE_DATA" interfaces were detected.
	*** This may be because your kernel version is newer than what is
	*** supported, or you are using a patched custom kernel with
	*** incompatible modifications.
	***
	*** ZFS Version: zfs-2.0.0-rc1
	*** Compatible Kernels: 3.10 - 5.9
```

However, when I go into `build/pde_data/` and run the following it works fine, and I can't gather any meaningful output from the log files:
```sh
make modules -C /<.....>/git/kernels/linux  M=/<.....>/src/zfs-test/build/pde_data
```

This is against `linux.git` commit `614cb5894306cfa2c7d9b6168182876ff5948735`.

**Comment by behlendorf** [comment]
@ckane could you point me at the relevant kernel commits.  I wasn't able to find them.

**Comment by ckane** [comment]
> @ckane could you point me at the relevant kernel commits. I wasn't able to find them.

Sure, here's the relevant change (I believe):
* https://github.com/torvalds/linux/commit/b9ed847b5ae69e0f2e685f9d53e2dd94c0db751e

**Comment by ckane** [comment]
> @ckane could you point me at the relevant kernel commits. I wasn't able to find them.

Sure, here are (what I think) are relevant changes:
* https://github.com/torvalds/linux/commit/1d6cd39293602e990b016994e51956eded35da7c (require `MODULE_LICENSE`)
* https://github.com/torvalds/linux/commit/d6d692fa21d3057edf457a764832077da8aa44d2 (GPL-incompatibility becomes fatal - not sure if this might be contributing to the PDE_DATA error, but I do see some complaints about this too in logs)

**Comment by behlendorf** [comment]
@ckane thanks for the links.  It looks like those changes were only merged yesterday which is why I hadn't seen them yet.  I've added another commit to the stack to handle the change from a warning to error.  Would you mind testing the updated stack in your environment.

**Comment by ckane** [comment]
Thanks, this appears to build on both the new kernel and older kernels for me.

File: config/kernel-blkdev.m4 [hunk 1]
```diff
@@ -154,42 +154,69 @@ AC_DEFUN([ZFS_AC_KERNEL_BLKDEV_INVALIDATE_BDEV], [
...
+		const char path[] = "/example/path";
+
+		bdev = lookup_bdev(path, FMODE_READ);
 	])
 ])
 
 AC_DEFUN([ZFS_AC_KERNEL_BLKDEV_LOOKUP_BDEV], [
-	AC_MSG_CHECKING([whether lookup_bdev() wants 1 arg])
-	ZFS_LINUX_TEST_RESULT_SYMBOL([lookup_bdev_1arg],
+	AC_MSG_CHECKING([whether lookup_vdev() want dev_t arg])
```
Comment (airminer) [inline config/kernel-blkdev.m4:193]:
```suggestion
	AC_MSG_CHECKING([whether lookup_bdev() want dev_t arg])
```
Typo.

**Comment by ckane** [comment]
I just built and am running off of your latest code now, too.

**Comment by rkitover** [comment]
@behlendorf I don't know very much about kernel APIs, but looks great to me!

One minor change I'd make is collapse the 1-arg and 2-arg lookup_bdev() blocks into one block, since they are identical except for one line.

**Comment by behlendorf** [comment]
@rkitover thanks for taking a look.  I had that same thought about `lookup_bdev()` and made that change originally, however in the end I felt it actually hurt readability so I opted to split it up.  I'd prefer to leave it as separate blocks even if it is a little redundant.

**Comment by rkitover** [comment]
Just wanted to say that I'm running this on gentoo with 5.11-rc1, no issues.

Thank you!

**Comment by averyfreeman** [comment]
I am running into this issue building 2.0.4-1 against 5.11.8 kernel. Here's the tail of `./configure`:

```
checking whether global_page_state enums are sane... yes
checking whether compile-time stack validation (objtool) is available... yes
checking whether STACK_FRAME_NON_STANDARD is defined... yes
checking whether PDE_DATA() is available... configure: error: 
        *** None of the expected "PDE_DATA" interfaces were detected.
        *** This may be because your kernel version is newer than what is
        *** supported, or you are using a patched custom kernel with
        *** incompatible modifications.
        ***
        *** ZFS Version: zfs-2.0.4-1
        *** Compatible Kernels: 3.10 - 5.11
```

Same message with clang-9 and gcc Debian 10.2.1 20210110 (I'm on bullseye/testing), and I also tried a clone of the master branch against 5.11.8.

I just tried with 5.11.6 kernel tarball and it passed `./configure` and is building right now, so something regarding PDA_DATA interface changed between 5.11.6 and 5.11.8.

Do you think I should file a separate bug report?

**Comment by ckane** [comment]
> I am running into this issue building 2.0.4-1 against 5.11.8 kernel. Here's the tail of `./configure`:
> 
> ```
> checking whether global_page_state enums are sane... yes
> checking whether compile-time stack validation (objtool) is available... yes
> checking whether STACK_FRAME_NON_STANDARD is defined... yes
> checking whether PDE_DATA() is available... configure: error: 
>         *** None of the expected "PDE_DATA" interfaces were detected.
>         *** This may be because your kernel version is newer than what is
>         *** supported, or you are using a patched custom kernel with
>         *** incompatible modifications.
>         ***
>         *** ZFS Version: zfs-2.0.4-1
>         *** Compatible Kernels: 3.10 - 5.11
> ```
> 
> Same message with clang-9 and gcc Debian 10.2.1 20210110 (I'm on bullseye/testing), and I also tried a clone of the master branch against 5.11.8.
> 
> I just tried with 5.11.6 kernel tarball and it passed `./configure` and is building right now, so something regarding PDA_DATA interface changed between 5.11.6 and 5.11.8.
> 
> Do you think I should file a separate bug report?

Yes, though first try building against the `master` branch to verify that this isn't already fixed in the next release, and try to include any relevant errors from `configure` that might have been spat into `./build/build.log*` files.

**Comment by rincebrain** [comment]
I also can't reproduce this with 5.11.8 on Debian bullseye (generated with make olddefconfig against config-5.10.0-4-amd64), so you should probably include your kernel .config somewhere along with config.log when you post a bug report.

**Comment by averyfreeman** [comment]
OK, thanks for the heads up, I do run zfs built-in kernels.

**Pull request closed without merging**